### PR TITLE
osd: optimize send_message to peers

### PR DESCRIPTION
**Pull request by majianpeng**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

**Comment by majianpeng** [comment]
send_message_osd_cluster is a hot fun which used by MOSDRepOP or MOSDECSubOpRead or MOSDSubOpWrite.  I did some optimization for this function.

**Comment by majianpeng** [comment]
jenkins retest this please

**Comment by majianpeng** [comment]
retest this please

**Comment by majianpeng** [comment]
@liewegas  please review ! Thanks

File: src/msg/async/AsyncConnection.h [hunk 1]
```diff
@@ -110,6 +110,7 @@ class AsyncConnection : public Connection {
   AsyncConnection(CephContext *cct, AsyncMessenger *m, DispatchQueue *q,
 		  Worker *w, bool is_msgr2, bool local);
   ~AsyncConnection() override;
+  bool unregisted = false;
```
Comment (liewegas) [inline src/msg/async/AsyncConnection.h]: nit: unregistered

File: src/osd/ECBackend.cc [hunk 2]
```diff
@@ -813,7 +813,7 @@ bool ECBackend::_handle_message(
     handle_sub_read(op->op.from, op->op, &(reply->op), _op->pg_trace);
     reply->trace = _op->pg_trace;
     get_parent()->send_message_osd_cluster(
```
Comment (liewegas) [inline src/osd/ECBackend.cc]: I think an even better improvement here would be ``op->op->get_connection()->send_message(...)``

**Comment by tchaikov** [comment]
jenkins test make check

File: src/osd/ECBackend.cc (hunk 2 above)
Comment (majianpeng) [reply src/osd/ECBackend.cc]: In ECBackend/ReplicatedBackend, it always use get_parent()->send_message_osd_cluster(). In order to be consistent, still unchanged. But we change PrimaryLogPG::send_message_osd_cluster into inline func to get the same aim.

**Comment by majianpeng** [comment]
jenkins test make check

**Comment by majianpeng** [comment]
retest this please

**Comment by tchaikov** [comment]
```
132/186 Test  #15: smoke.sh ................................***Failed  176.97 sec
```

jenkins test make check

**Comment by majianpeng** [comment]
@tchaikov .  why always failed? because my pr or?

**Comment by tchaikov** [comment]
```
136/183 Test  #15: smoke.sh ................................***Failed  483.46 sec
```

jenkins test make check arm64

**Comment by tchaikov** [comment]
@majianpeng no idea. probably you could try to run `smoke.sh` in your testbed?

**Comment by majianpeng** [comment]
@tchaikov .i'll try in my host

**Comment by majianpeng** [comment]
@tchaikov , in my test, w/o this pr smoke.sh still failed.

**Comment by majianpeng** [comment]
@tchaikov. omit my previous message. this bug caused by 97716c1.

**Comment by liewegas** [comment]
Seeing lots of failures like this:
<pre>
2019-10-28T03:23:17.527 INFO:tasks.ceph.osd.4.smithi203.stderr:*** Caught signal (Segmentation fault) **
2019-10-28T03:23:17.528 INFO:tasks.ceph.osd.4.smithi203.stderr: in thread 7f9f22622700 thread_name:safe_timer
2019-10-28T03:23:17.530 INFO:tasks.ceph.osd.4.smithi203.stderr: ceph version 15.0.0-6532-g37ccb12 (37ccb123c3fc8f973766084a4e44b4fbf16df8bf) octopus (dev)
2019-10-28T03:23:17.530 INFO:tasks.ceph.osd.4.smithi203.stderr: 1: (()+0xf630) [0x7f9f2ee78630]
2019-10-28T03:23:17.530 INFO:tasks.ceph.osd.4.smithi203.stderr: 2: (AsyncConnection::send_message(Message*)+0x201) [0x55e3269087b1]
2019-10-28T03:23:17.530 INFO:tasks.ceph.osd.4.smithi203.stderr: 3: (OSDService::send_message_osd_cluster(std::vector<std::pair<int, Message*>, std::allocator<std::pair<int, Message*> > >&, unsigned int)+0xec) [0x55e32601ff6c]
2019-10-28T03:23:17.530 INFO:tasks.ceph.osd.4.smithi203.stderr: 4: (PG::scrub_reserve_replicas()+0x2e2) [0x55e3260cbff2]
2019-10-28T03:23:17.531 INFO:tasks.ceph.osd.4.smithi203.stderr: 5: (PG::sched_scrub()+0x592) [0x55e3260cc982]
2019-10-28T03:23:17.531 INFO:tasks.ceph.osd.4.smithi203.stderr: 6: (OSD::sched_scrub()+0x4fc) [0x55e32602443c]
2019-10-28T03:23:17.531 INFO:tasks.ceph.osd.4.smithi203.stderr: 7: (OSD::tick_without_osd_lock()+0x650) [0x55e3260323c0]
2019-10-28T03:23:17.531 INFO:tasks.ceph.osd.4.smithi203.stderr: 8: (Context::complete(int)+0x9) [0x55e3260624a9]
2019-10-28T03:23:17.531 INFO:tasks.ceph.osd.4.smithi203.stderr: 9: (SafeTimer::timer_thread()+0x1a8) [0x55e32660b268]
2019-10-28T03:23:17.531 INFO:tasks.ceph.osd.4.smithi203.stderr: 10: (SafeTimerThread::entry()+0xd) [0x55e32660c65d]
2019-10-28T03:23:17.531 INFO:tasks.ceph.osd.4.smithi203.stderr: 11: (()+0x7ea5) [0x7f9f2ee70ea5]
2019-10-28T03:23:17.532 INFO:tasks.ceph.osd.4.smithi203.stderr: 12: (clone()+0x6d) [0x7f9f2dd348cd]
2019-10-28T03:23:17.532 INFO:tasks.ceph.osd.4.smithi203.stderr:2019-10-28T03:23:17.533+0000 7f9f22622700 -1 *** Caught signal (Segmentation fault) **
</pre>
/a/sage-2019-10-28_02:58:26-rados-wip-sage-testing-2019-10-27-1006-distro-basic-smithi/4450173

lots of others (with coredumps etc) in that same test run.

**Comment by majianpeng** [comment]
retest this please

**Comment by majianpeng** [comment]
@liewegas . i found the reason and fixed.

**Comment by liewegas** [comment]
```
2019-11-05T20:52:33.898 INFO:tasks.ceph.osd.4.smithi134.stderr: ceph version 15.0.0-6875-g6fbd5b8 (6fbd5b8f6235494ccd5d67cf76110c824622eacb) octopus (dev)
2019-11-05T20:52:33.898 INFO:tasks.ceph.osd.4.smithi134.stderr: 1: (()+0x12890) [0x7fa4a36b4890]
2019-11-05T20:52:33.898 INFO:tasks.ceph.osd.4.smithi134.stderr: 2: (AsyncConnection::send_message(Message*)+0x492) [0x5638cf4c88d2]
2019-11-05T20:52:33.899 INFO:tasks.ceph.osd.4.smithi134.stderr: 3: (OSDService::send_message_osd_cluster(std::vector<std::pair<int, Message*>, std::allocator<std::pair<int, Message*> > >&, unsigned int)+0x13c) [0x5638ceb492cc]
2019-11-05T20:52:33.899 INFO:tasks.ceph.osd.4.smithi134.stderr: 4: (ECBackend::do_read_op(ECBackend::ReadOp&)+0xca1) [0x5638ceebf821]
2019-11-05T20:52:33.899 INFO:tasks.ceph.osd.4.smithi134.stderr: 5: (ECBackend::start_read_op(int, std::map<hobject_t, std::set<int, std::less<int>, std::allocator<int> >, std::less<hobject_t>, std::allocator<std::pair<hobject_t const, std::set<int, std::less<int>, std::allocator<int> > > > >&, std::map<hobject_t, ECBackend::read_request_t, std::less<hobject_t>, std::allocator<std::pair<hobject_t const, ECBackend::read_request_t> > >&, boost::intrusive_ptr<OpRequest>, bool, bool)+0x465) [0x5638ceec0025]
2019-11-05T20:52:33.899 INFO:tasks.ceph.osd.4.smithi134.stderr: 6: (ECBackend::objects_read_and_reconstruct(std::map<hobject_t, std::__cxx11::list<boost::tuples::tuple<unsigned long, unsigned long, unsigned int, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type>, std::allocator<boost::tuples::tuple<unsigned long, unsigned long, unsigned int, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type> > >, std::less<hobject_t>, std::allocator<std::pair<hobject_t const, std::__cxx11::list<boost::tuples::tuple<unsigned long, unsigned long, unsigned int, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type>, std::allocator<boost::tuples::tuple<unsigned long, unsigned long, unsigned int, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type> > > > > > const&, bool, std::unique_ptr<GenContext<std::map<hobject_t, std::pair<int, interval_map<unsigned long, ceph::buffer::v14_2_0::list, bl_split_merge> >, std::less<hobject_t>, std::allocator<std::pair<hobject_t const, std::pair<int, interval_map<unsigned long, ceph::buffer::v14_2_0::list, bl_split_merge> > > > >&&>, std::default_delete<GenContext<std::map<hobject_t, std::pair<int, interval_map<unsigned long, ceph::buffer::v14_2_0::list, bl_split_merge> >, std::less<hobject_t>, std::allocator<std::pair<hobject_t const, std::pair<int, interval_map<unsigned long, ceph::buffer::v14_2_0::list, bl_split_merge> > > > >&&> > >&&)+0x7f2) [0x5638ceec14a2]
2019-11-05T20:52:33.899 INFO:tasks.ceph.osd.4.smithi134.stderr: 7: (ECBackend::objects_read_async(hobject_t const&, std::__cxx11::list<std::pair<boost::tuples::tuple<unsigned long, unsigned long, unsigned int, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type>, std::pair<ceph::buffer::v14_2_0::list*, Context*> >, std::allocator<std::pair<boost::tuples::tuple<unsigned long, unsigned long, unsigned int, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type>, std::pair<ceph::buffer::v14_2_0::list*, Context*> > > > const&, Context*, bool)+0x369) [0x5638ceec1b09]
2019-11-05T20:52:33.899 INFO:tasks.ceph.osd.4.smithi134.stderr: 8: (PrimaryLogPG::OpContext::start_async_reads(PrimaryLogPG*)+0xe8) [0x5638cec46048]
2019-11-05T20:52:33.900 INFO:tasks.ceph.osd.4.smithi134.stderr: 9: (PrimaryLogPG::execute_ctx(PrimaryLogPG::OpContext*)+0x3ea) [0x5638cecc15ca]
2019-11-05T20:52:33.900 INFO:tasks.ceph.osd.4.smithi134.stderr: 10: (PrimaryLogPG::do_op(boost::intrusive_ptr<OpRequest>&)+0x2f5d) [0x5638cecc5c0d]
2019-11-05T20:52:33.900 INFO:tasks.ceph.osd.4.smithi134.stderr: 11: (PrimaryLogPG::do_request(boost::intrusive_ptr<OpRequest>&, ThreadPool::TPHandle&)+0xcf1) [0x5638cecd1b81]
2019-11-05T20:52:33.902 INFO:tasks.ceph.osd.4.smithi134.stderr: 12: (OSD::dequeue_op(boost::intrusive_ptr<PG>, boost::intrusive_ptr<OpRequest>, ThreadPool::TPHandle&)+0x17b) [0x5638ceb57a8b]
2019-11-05T20:52:33.902 INFO:tasks.ceph.osd.4.smithi134.stderr: 13: (ceph::osd::scheduler::PGOpItem::run(OSD*, OSDShard*, boost::intrusive_ptr<PG>&, ThreadPool::TPHandle&)+0x67) [0x5638cedb26c7]
2019-11-05T20:52:33.902 INFO:tasks.ceph.osd.4.smithi134.stderr: 14: (OSD::ShardedOpWQ::_process(unsigned int, ceph::heartbeat_handle_d*)+0x90c) [0x5638ceb751fc]
2019-11-05T20:52:33.902 INFO:tasks.ceph.osd.4.smithi134.stderr: 15: (ShardedThreadPool::shardedthreadpool_worker(unsigned int)+0x4ac) [0x5638cf1ae8dc]
2019-11-05T20:52:33.902 INFO:tasks.ceph.osd.4.smithi134.stderr: 16: (ShardedThreadPool::WorkThreadSharded::entry()+0x10) [0x5638cf1b1a50]
2019-11-05T20:52:33.902 INFO:tasks.ceph.osd.4.smithi134.stderr: 17: (()+0x76db) [0x7fa4a36a96db]
2019-11-05T20:52:33.902 INFO:tasks.ceph.osd.4.smithi134.stderr: 18: (clone()+0x3f) [0x7fa4a244988f]
2019-11-05T20:52:33.903 INFO:tasks.ceph.osd.4.smithi134.stderr:2019-11-05T20:52:33.897+0000 7fa47d75c700 -1 *** Caught signal (Segmentation fault) **
```
/a/sage-2019-11-05_17:51:43-rados-wip-sage-testing-2019-11-05-0856-distro-basic-smithi/4475123

lots of similar failures in this run

**Comment by majianpeng** [comment]
@liewegas . Bug fixed and I hope this can pass all the tests.  Thanks!

**Comment by majianpeng** [comment]
@liewegas . ping

**Comment by liewegas** [comment]
yep, it's marked... should get it on my next run

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2019-11-17_01:29:22-rados-wip-kefu-testing-2019-11-17-0135-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-11-19_00:33:33-rados-wip-kefu-testing-2019-11-17-0135-distro-basic-mira/

test failures are either tracked or fixed by

- https://tracker.ceph.com/issues/42876
- #31689

**Pull request merged**